python/lsst/afw/table/catalogMatches.py

```diff
@@ -0,0 +1,164 @@
+# 
+# LSST Data Management System
+# Copyright 2008-2015 AURA/LSST.
+# 
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+# 
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+# 
+# You should have received a copy of the LSST License Statement and 
+# the GNU General Public License along with this program.  If not, 
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+from __future__ import absolute_import
+
+import os.path
+
+from .tableLib import BaseCatalog, SimpleCatalog, SourceCatalog, SimpleTable, SourceTable, \
+                      Schema, ReferenceMatch
+from lsst.utils import getPackageDir
+
+__all__ = ["copySchema", "copyCatalog", "matchesToCatalog", "matchesFromCatalog"]
+
+def copySchema(schema, target, targetPrefix=None, sourcePrefix=None):
+    """Return a deep copy the provided schema
+
+    If sourcePrefix is set, only copy those keys that have that prefix.
+    If targetPrefix is set, add that to the key name.
+    """
+    existing = set(target.getNames())
+    for keyName in schema.getNames():
+        if sourcePrefix is not None:
+            if not keyName.startswith(sourcePrefix):
+                continue
+            keyNameFix = keyName[len(sourcePrefix):]
+        else:
+            keyNameFix = keyName
+        if keyNameFix in existing:
+            continue
+        field = schema.find(keyName).field
+        typeStr = field.getTypeString()
+        fieldDoc = field.getDoc()
+        fieldUnits = field.getUnits()
+        if typeStr in ("ArrayF", "ArrayD", "ArrayI", "CovF", "CovD"):
+            fieldSize = field.getSize()
+        else:
+            fieldSize = None
+
+        target.addField((targetPrefix if targetPrefix is not None else "") + keyNameFix,
+                        type=typeStr, doc=fieldDoc, units=fieldUnits, size=fieldSize)
+    return target
+
+def copyCatalog(catalog, target, sourceSchema=None, targetPrefix=None, sourcePrefix=None):
```

Comment:
Similar comments apply here as to `copySchema()`, above.

* Please clarify whether callers can expect the `target` they supply to be modified or if they need to catch the return value;
* This could could be simplified quite a bit by using a `SchemaMapper`;
* Even if you don't want to rewrite to use the `SchemaMapper` approach, there's some duplication between these functions in the handling of `sourcePrefix` that you could consider trying to eliminate.